### PR TITLE
RPyC update docs and make it easier to use

### PR DESCRIPTION
**Pull request by dwalton76**
For issue #677

### CODE REVIEW

**Comment by dwalton76** [comment]
See preview [here](http://rst.ninjs.org/#KioqKioqKioqKioqKioKUlB5QyBvbiBldjNkZXYKKioqKioqKioqKioqKioKCmBSUHlDXyA8aHR0cHM6Ly9ycHljLnJlYWR0aGVkb2NzLmlvL2VuL2xhdGVzdC8+YF8gKHByb25vdW5jZWQgYXMgYXJlLXBpZS1zZWUpLCBvciBSZW1vdGUgUHl0aG9uIENhbGwsIGlzIGEgdHJhbnNwYXJlbnQKcHl0aG9uIGxpYnJhcnkgZm9yIHN5bW1ldHJpY2FsIHJlbW90ZSBwcm9jZWR1cmUgY2FsbHMsIGNsdXN0ZXJpbmcgYW5kCmRpc3RyaWJ1dGVkLWNvbXB1dGluZy4gUlB5QyBtYWtlcyB1c2Ugb2Ygb2JqZWN0LXByb3h5aW5nLCBhIHRlY2huaXF1ZSB0aGF0CmVtcGxveXMgcHl0aG9u4oCZcyBkeW5hbWljIG5hdHVyZSwgdG8gb3ZlcmNvbWUgdGhlIHBoeXNpY2FsIGJvdW5kYXJpZXMgYmV0d2Vlbgpwcm9jZXNzZXMgYW5kIGNvbXB1dGVycywgc28gdGhhdCByZW1vdGUgb2JqZWN0cyBjYW4gYmUgbWFuaXB1bGF0ZWQgYXMgaWYgdGhleQp3ZXJlIGxvY2FsLgoKRm9yIGV2M2RldiwgUlB5QyBpcyBtb3N0IG9mdGVuIHVzZWQgZm9yCgoqIHJvYm90cyB0aGF0IGludm9sdmUgbW9yZSB0aGFuIG9uZSBFVjMgKGkuZS4gZGFpc3kgY2hhaW5pbmcpCiogcm9ib3RzIHRoYXQgcGVyZm9ybSBzb21lIENQVSBpbnRlbnNpdmUgdGFzayAoZXg6IFJ1YmlrJ3MgY3ViZSBzb2x2ZXIpIHdoZXJlIHlvdQogIHdpc2ggdG8gcnVuIHRoZSBDUFUgaW50ZW5zaXZlIHBhcnQgb24geW91ciBkZXNrdG9wIFBDCgpOZXR3b3JraW5nCj09PT09PT09PT0KWW91IHdpbGwgbmVlZCB0byBzZXR1cCBuZXR3b3JraW5nIG9uIHlvdXIgZXYzZGV2IGRldmljZShzKS4gVGhlCmBldjNkZXYgbmV0d29ya2luZyBkb2N1bWVudGF0aW9uIDxodHRwczovL3d3dy5ldjNkZXYub3JnL2RvY3MvbmV0d29ya2luZy8+YF8gc2hvdWxkIGdldAp5b3UgdXAgYW5kIHJ1bm5pbmcgaW4gdGVybXMgb2YgbmV0d29ya2luZyBjb25uZWN0aXZpdHkuCgoKSW5zdGFsbAo9PT09PT09CgoxLiBSUHlDIGlzIGluc3RhbGxlZCBvbiBldjNkZXYgYnV0IHdlIG5lZWQgdG8gY3JlYXRlIGEgc2VydmljZSB0aGF0IGxhdW5jaGVzCiAgIGBgcnB5Y19jbGFzc2ljLnB5YGAgYXQgYm9vdHVwLiBDdXQtbi1wYXN0ZSB0aGUgZm9sbG93aW5nIGNvbW1hbmRzIG9uIHRoZQogICBldjNkZXYgZGV2aWNlKHMpIHRoYXQgeW91IHdpc2ggdG8gY29udHJvbCB2aWEgUlB5Qy4KCiAgIC4uIGNvZGUtYmxvY2s6OiBzaGVsbAoKICAgICAgZWNobyAiW1VuaXRdCiAgICAgIERlc2NyaXB0aW9uPVJQeUMgQ2xhc3NpYyBTZXJ2aWNlCiAgICAgIEFmdGVyPW11bHRpLXVzZXIudGFyZ2V0CgogICAgICBbU2VydmljZV0KICAgICAgVHlwZT1zaW1wbGUKICAgICAgRXhlY1N0YXJ0PS91c3IvYmluL3JweWNfY2xhc3NpYy5weQoKICAgICAgW0luc3RhbGxdCiAgICAgIFdhbnRlZEJ5PW11bHRpLXVzZXIudGFyZ2V0IiA+IHJweWMtY2xhc3NpYy5zZXJ2aWNlCgogICAgICBzdWRvIGNwIHJweWMtY2xhc3NpYy5zZXJ2aWNlIC9saWIvc3lzdGVtZC9zeXN0ZW0vCiAgICAgIHN1ZG8gc3lzdGVtY3RsIGRhZW1vbi1yZWxvYWQKICAgICAgc3VkbyBzeXN0ZW1jdGwgZW5hYmxlIHJweWMtY2xhc3NpYy5zZXJ2aWNlCiAgICAgIHN1ZG8gc3lzdGVtY3RsIHN0YXJ0IHJweWMtY2xhc3NpYy5zZXJ2aWNlCgoKMi4gSWYgeW91IHdpbGwgYmUgdXNpbmcgYW4gZXYzZGV2IGRldmljZSB0byBjb250cm9sIGFub3RoZXIgZXYzZGV2IGRldmljZSB5b3UKICAgY2FuIHNraXAgdGhpcyBzdGVwLiAgSWYgeW91IHdpbGwgYmUgdXNpbmcgeW91ciBkZXNrdG9wIFBDIHRvIGNvbnRyb2wgYW4gZXYzZGV2CiAgIGRldmljZSB5b3UgbXVzdCBpbnN0YWxsIFJQeUMgb24geW91ciBkZXNrdG9wIFBDLiBIb3cgeW91IGluc3RhbGwgUlB5QyBkZXBlbmRzCiAgIG9uIHlvdXIgb3BlcmF0aW5nIHN5c3RlbS4gRm9yIExpbnV4IHlvdSBzaG91bGQgYmUgYWJsZSB0byBkbzoKCiAgIC4uIGNvZGUtYmxvY2s6OiBzaGVsbAoKICAgICAgIHN1ZG8gYXB0LWdldCBpbnN0YWxsIHB5dGhvbjMtcnB5YwoKICAgRm9yIFdpbmRvd3MgdGhlcmUgaXMgYSB3aW4zMiBpbnN0YWxsZXIgb24gdGhlIHByb2plY3QncyBgc291cmNlZm9yZ2UgcGFnZWBfLgogICBBbHNvLCBoYXZlIGEgbG9vayBhdCB0aGUgYERvd25sb2FkIGFuZCBJbnN0YWxsYF8gcGFnZSBvbiB0aGVpciBzaXRlLgoKRXhhbXBsZQo9PT09PT09ClRoZSBmb2xsb3dpbmcgcHl0aG9uIHNjcmlwdCBzaG91bGQgbWFrZSBhIGxhcmdlIG1vdG9yIGNvbm5lY3RlZCB0byBgYE9VVFBVVF9BYGAKcnVuIHdoaWxlIHRoZSB0b3VjaCBzZW5zb3Igb24gYGBJTlBVVF8xYGAgaXMgcHJlc3NlZC4KCiAgIC4uIGNvZGUtYmxvY2s6OiBweQoKICAgICAgIGltcG9ydCBycHljCgogICAgICAgIyBDcmVhdGUgYSBSUHlDIGNvbm5lY3Rpb24gdG8gdGhlIHJlbW90ZSBldjNkZXYgZGV2aWNlLgogICAgICAgIyBVc2UgdGhlIGhvc3RuYW1lIG9yIElQIGFkZHJlc3Mgb2YgdGhlIGV2M2RldiBkZXZpY2UuCiAgICAgICAjIElmIHRoaXMgZmFpbHMsIHZlcmlmeSB5b3VyIElQIGNvbm5lY3RpdnR5IHZpYSBgYHBpbmcgWC5YLlguWGBgCiAgICAgICBjb25uID0gcnB5Yy5jbGFzc2ljLmNvbm5lY3QoJ1guWC5YLlgnKQoKICAgICAgICMgaW1wb3J0IGV2M2RldjIgb24gdGhlIHJlbW90ZSBldjNkZXYgZGV2aWNlCiAgICAgICBldjNkZXYyX21vdG9yID0gY29ubi5tb2R1bGVzWydldjNkZXYyLm1vdG9yJ10KICAgICAgIGV2M2RldjJfc2Vuc29yID0gY29ubi5tb2R1bGVzWydldjNkZXYyLnNlbnNvciddCiAgICAgICBldjNkZXYyX3NlbnNvcl9sZWdvID0gY29ubi5tb2R1bGVzWydldjNkZXYyLnNlbnNvci5sZWdvJ10KCiAgICAgICAjIFVzZSB0aGUgTGFyZ2VNb3RvciBhbmQgVG91Y2hTZW5zb3Igb24gdGhlIHJlbW90ZSBldjNkZXYgZGV2aWNlCiAgICAgICBtb3RvciA9IGV2M2RldjJfbW90b3IuTGFyZ2VNb3RvcihldjNkZXYyX21vdG9yLk9VVFBVVF9BKQogICAgICAgdHMgPSBldjNkZXYyX3NlbnNvcl9sZWdvLlRvdWNoU2Vuc29yKGV2M2RldjJfc2Vuc29yLklOUFVUXzEpCgogICAgICAgIyBJZiB0aGUgVG91Y2hTZW5zb3IgaXMgcHJlc3NlZCwgcnVuIHRoZSBtb3RvcgogICAgICAgd2hpbGUgVHJ1ZToKICAgICAgICAgICB0cy53YWl0X2Zvcl9wcmVzc2VkKCkKICAgICAgICAgICBtb3Rvci5ydW5fZm9yZXZlcihzcGVlZF9zcD0yMDApCgogICAgICAgICAgIHRzLndhaXRfZm9yX3JlbGVhc2VkKCkKICAgICAgICAgICBtb3Rvci5zdG9wKCkKCgpQcm9zCj09PT0KKiBUaGUgUlB5QyBzZXJ2ZXIgaXMgbGlnaHR3ZWlnaHQgYW5kIG9ubHkgcmVxdWlyZXMgYW4gSVAgY29ubmVjdGlvbiAobm8gc3NoIHJlcXVpcmVkKS4KKiBTb21lIHJvYm90cyBtYXkgbmVlZCBtdWNoIG1vcmUgY29tcHV0YXRpb25hbCBwb3dlciB0aGFuIGFuIEVWMyBjYW4gZ2l2ZQogIHlvdS4gQSBub3RhYmxlIGV4YW1wbGUgaXMgdGhlIFJ1YmlrJ3MgY3ViZSBzb2x2ZXIuIFRoZXJlIGlzIGFuIGFsZ29yaXRobSB0aGF0CiAgcHJvdmlkZXMgYW4gYWxtb3N0IG9wdGltYWwgc29sdXRpb24gKGluIHRlcm1zIG9mIG51bWJlciBvZiBjdWJlIHJvdGF0aW9ucyksIGJ1dAogIGl0IHRha2VzIG1vcmUgUkFNIHRoYW4gaXMgYXZhaWxhYmxlIG9uIEVWMy4gV2l0aCBSUHlDLCB5b3UgY291bGQgcnVuIHRoZQogIGhlYXZ5LWR1dHkgY29tcHV0YXRpb25zIG9uIHlvdXIgZGVza3RvcC4KCkNvbnMKPT09PQpUaGUgbW9zdCBvYnZpb3VzICpkaXNhZHZhbnRhZ2UqIGlzIGxhdGVuY3kgaW50cm9kdWNlZCBieSBuZXR3b3JrIGNvbm5lY3Rpb24uClRoaXMgbWF5IGJlIGEgc2hvdyBzdG9wcGVyIGZvciByb2JvdHMgd2hlcmUgcmVhY3Rpb24gc3BlZWQgaXMgZXNzZW50aWFsLgoKUmVmZXJlbmNlcwo9PT09PT09PT09CiogYFJQeUMgPGh0dHA6Ly9ycHljLnJlYWR0aGVkb2NzLmlvLz5gXwoqIGBzb3VyY2Vmb3JnZSBwYWdlIDxodHRwOi8vc291cmNlZm9yZ2UubmV0L3Byb2plY3RzL3JweWMvZmlsZXMvbWFpbj5gXwoqIGBEb3dubG9hZCBhbmQgSW5zdGFsbCA8aHR0cDovL3JweWMucmVhZHRoZWRvY3MuaW8vZW4vbGF0ZXN0L2luc3RhbGwuaHRtbD5gXwoqIGBjb25uZWN0IHdpdGggU1NIIDxodHRwOi8vd3d3LmV2M2Rldi5vcmcvZG9jcy90dXRvcmlhbHMvY29ubmVjdGluZy10by1ldjNkZXYtd2l0aC1zc2gvPmBfCg==
)

File: docs/rpyc.rst [hunk 1]
```diff
@@ -1,87 +1,114 @@
-Working with ev3dev remotely using RPyC
-=======================================
+**************
+RPyC on ev3dev
+**************
 
-RPyC_ (pronounced as are-pie-see), or Remote Python Call, is a transparent
+`RPyC_ <https://rpyc.readthedocs.io/en/latest/>`_ (pronounced as are-pie-see), or Remote Python Call, is a transparent
 python library for symmetrical remote procedure calls, clustering and
 distributed-computing. RPyC makes use of object-proxying, a technique that
```
Comment (WasabiFan) [inline docs/rpyc.rst]: This is copied from the RPyC homepage; can we add attribution? Additionally, I think it's easy for someone to be scared off if this is the first thing they read. I'd say we should explain it in simpler terms first then just quote their description if needed.

File: docs/rpyc.rst [hunk 2]
```diff
@@ -1,87 +1,114 @@
...
-   following command at the command prompt (after you `connect with SSH`_):
+For ev3dev, RPyC is most often used for:
+
+* robots that involve more than one EV3 (i.e. daisy chaining)
+* robots that perform some CPU intensive task (ex: Rubik's cube solver) where you
+  wish to run the CPU intensive part on your desktop PC
+
+Networking
+==========
+You will need to setup networking on your ev3dev device(s). The
```
Comment (WasabiFan) [inline docs/rpyc.rst]: Is the idea that they should have a network connection _between their host and client_ here? If so, maybe mention that and highlight some of the types that would work:

- Two EV3s on the same Wi-Fi network
- A host PC or EV3 connected to a client EV3
- A bluetooth connection between two EV3s

File: docs/rpyc.rst [hunk 3]
```diff
@@ -1,87 +1,114 @@
...
 
-RPyC_ (pronounced as are-pie-see), or Remote Python Call, is a transparent
+`RPyC_ <https://rpyc.readthedocs.io/en/latest/>`_ (pronounced as are-pie-see), or Remote Python Call, is a transparent
 python library for symmetrical remote procedure calls, clustering and
 distributed-computing. RPyC makes use of object-proxying, a technique that
 employs python’s dynamic nature, to overcome the physical boundaries between
 processes and computers, so that remote objects can be manipulated as if they
-were local. Here are simple steps you need to follow in order to install and
-use RPyC with ev3dev:
+were local.
```
Comment (WasabiFan) [inline docs/rpyc.rst]: I'd suggest we have a short section here which explains what you can actually do with RPyC. As in -- how it looks in code to use it.

File: docs/rpyc.rst [hunk 4]
```diff
@@ -1,87 +1,114 @@
...
-   and make the file executable:
+2. If you will be using an ev3dev device to control another ev3dev device you
+   can skip this step.  If you will be using your desktop PC to control an ev3dev
+   device you must install RPyC on your desktop PC. How you install RPyC depends
+   on your operating system. For Linux you should be able to do:
 
    .. code-block:: shell
 
-      chmod +x rpyc_server.sh
+       sudo apt-get install python3-rpyc
```
Comment (WasabiFan) [inline docs/rpyc.rst:62]: Why not have them install via `pip` so it's the same between platforms? (although, the `pip` vs `pip3` distinction might be important). Or point them at the documentation for RPyC?

File: docs/rpyc.rst [hunk 5]
```diff
@@ -1,87 +1,114 @@
...
+
+   For Windows there is a win32 installer on the project's `sourceforge page`_.
+   Also, have a look at the `Download and Install`_ page on their site.
+
+Example
+=======
+The following python script should make a large motor connected to ``OUTPUT_A``
```
Comment (WasabiFan) [inline docs/rpyc.rst]: I think more description is needed about which device this is run on and what it will do (operate on the remote device).

File: docs/rpyc.rst [hunk 6]
```diff
@@ -1,87 +1,114 @@
...
+* Some robots may need much more computational power than an EV3 can give
+  you. A notable example is the Rubik's cube solver. There is an algorithm that
+  provides an almost optimal solution (in terms of number of cube rotations), but
+  it takes more RAM than is available on EV3. With RPyC, you could run the
   heavy-duty computations on your desktop.
 
+Cons
+====
 The most obvious *disadvantage* is latency introduced by network connection.
 This may be a show stopper for robots where reaction speed is essential.
```
Comment (WasabiFan) [inline docs/rpyc.rst]: Also requires full python3 rather than micropython

File: docs/rpyc.rst [hunk 7]
```diff
@@ -1,87 +1,114 @@
...
 
-      INFO:SLAVE/18812:server started on [0.0.0.0]:18812
+       # Create a RPyC connection to the remote ev3dev device.
+       # Use the hostname or IP address of the ev3dev device.
+       # If this fails, verify your IP connectivty via ``ping X.X.X.X``
+       conn = rpyc.classic.connect('X.X.X.X')
 
-   and keep running.
+       # import ev3dev2 on the remote ev3dev device
+       ev3dev2_motor = conn.modules['ev3dev2.motor']
```
Comment (WasabiFan) [inline docs/rpyc.rst:83]: have you tested this?

File: docs/rpyc.rst (hunk 1 above)
Comment (dwalton76) [reply docs/rpyc.rst]: yeah I think this whole paragraph is a bit intimidating. Will chop it with some short explanation of why someone would want to use it either ev3dev.

File: docs/rpyc.rst (hunk 4 above)
Comment (dwalton76) [reply docs/rpyc.rst:62]: Wasn’t David encouraging apt install over pip? I was going based off of that.

File: docs/rpyc.rst (hunk 7 above)
Comment (dwalton76) [reply docs/rpyc.rst:83]: yep, it works like a champ

File: docs/rpyc.rst (hunk 6 above)
Comment (dwalton76) [reply docs/rpyc.rst]: You know there is part of me that thinks about everything we support via micropython now and how much faster it starts...should we just drop the RPyC angle and push MQTT as “the” daisy chain solution? It is would be less confusing in that there would only be one choice AND it would work for both python3 and micropython.

File: docs/rpyc.rst (hunk 4 above)
Comment (WasabiFan) [reply docs/rpyc.rst:62]: I can't remember where that was or the context behind it so I'm not sure. Using apt is also OK as long as we have a reasonable route for Mac/Windows users.

File: docs/rpyc.rst (hunk 6 above)
Comment (WasabiFan) [reply docs/rpyc.rst]: I wouldn't be opposed to that. I do like the MQTT approach if we figure out a good way to package it up/describe it. I started replying to the other thread but got distracted and will finish.

Comment (dwalton76) [reply docs/rpyc.rst]: yeah lets see what shakes out of this mqtt/daisy-chain brainstorming.
